충남대 BE_이은경 5주차 과제 (1단계)

pkyung opened the pull request:
## 과제 진행 사항
### 카카오 로그인
- [x] 카카오 로그인을 통해 인가 코드를 받는다.
- [x] 액세스 토큰을 추출한다. 
- [x] scope 없을 경우 403 에러 반환한다.

pkyung committed:
Chore: 4주차 코드 옭김

pkyung committed:
Chore: application.properties 삭제

pkyung committed:
Docs: 기능 목록 작성 - step1

pkyung committed:
Feat: 카카오 로그인 엑세스 토큰 추출

1. kakaoProperties로 중요한 정보를 관리한다.
2. Properties를 application에 @EnableConfigurationProperties를 통해
   등록한다.
3. RestTemplate을 빈으로 등록한다.

pkyung committed:
Feat: scope 없을 경우 403 에러 반환

1. reponseBody에 talk_message가 없을 경우 403 에러를 반환한다.
2. 있다면 응답값을 반환한다.

pkyung committed:
Docs: 기능 목록 구현 - step1

pkyung committed:
Test: getAccessToken 메서드 테스트

pkyung committed:
Merge branch 'pkyung' into main

ksy90101 commented on src/main/java/gift/controller/OAuthController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @GetMapping("/")
+    public ResponseEntity<String> getAccessToken(@RequestParam("code") String code) {
```
토큰은 얻은 다음에, 회원가입 및 로그인이 되어야 할거 같은데 그 로직이 없는거 같아요!?

pkyung replied:
회원가입과 로그인 과정을 넣도록 하겠습니다.

ksy90101 reviewed (changes requested):
고생하셨습니다!!

ksy90101 commented on src/main/java/gift/controller/OAuthController.java:
```diff
@@ -0,0 +1,53 @@
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED_VALUE);
+
+        LinkedMultiValueMap<String, String> body = new LinkedMultiValueMap<>();
```
추상회된 Map interface로 타입을 쓰는게 더 어울리지 않을까요?

pkyung replied:
네 hashmap으로 수정하겠습니다.

pkyung replied:
multiValueMap으로 수정하겠습니다.

pkyung committed:
Feat: OAuth2 로그인, 회원가입 구현

1. 얻은 accessToken으로 사용자 프로필을 요청했다.
2. 사용자 프로필의 Id 앞에 kakao 키워드를 붙여 email 로 멤버를 자동
   가입시켰다.
3. 멤버가 이미 가입되어있거나 자동 가입된 경우 토큰을 반환하도록
   구현했다.

pkyung committed:
Test: 카카오 로그인 테스트

pkyung committed:
Refactor: api component로 api 요청 분리

pkyung committed:
Refactor: member에 accessToken 추가

member 객체에 accessToken을 추가하여 저장했다.

ksy90101 commented on src/main/java/gift/component/KakaoApiComponent.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public String getAccessToken(String code) {
+        String url = "https://kauth.kakao.com/oauth/token";
```
이건 상수가 어울리지 않을까...
변하지 않는 값일테니깐요.

ksy90101 commented on src/main/java/gift/component/KakaoApiComponent.java:
```diff
@@ -0,0 +1,72 @@
+        this.objectMapper = objectMapper;
+    }
+
+    public String getAccessToken(String code) {
```
해당 메서드가 너무 많은일을 하고 있는거 같아요.
역할ㄹ을 메서드로 분리해보면 어떨가 생각이 듭니다.

분리하는 관점이 어려우면 DM주시면 허들로 이야기 한번 드릴게요~

ksy90101 reviewed (approved):
어려운 요구 잘 충족해주셔서 감사합니다~
고생많으셨어요!
마무리할게요~
